### PR TITLE
Channels asgi setup

### PR DESCRIPTION
**Pull request by diegiesskanne**
Feel free to ask questions and tell me when there are some problems. I fix the pipeline errors tomorrow. I can squash the commits on merge later

### CODE REVIEW

File: embark/embark/consumers.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+        # observer for status messages
+        source_stream.pipe(
+            ops.map(lambda s: re.sub(color_pattern, '', s)),
+            ops.filter(lambda s: self.process_line(s, status_pattern)),
+            ops.map(lambda a: a.split("- ")),
+            ops.map(lambda t: t[1]),
+            ops.map(lambda b: b.split(" "))
+        ).subscribe(
+            lambda x: self.update_status(x)
```
Comment (MaximilianWagner) [inline embark/embark/consumers.py:157]: Maybe you could state shortly how the stream is processed

File: embark/embark/settings.py [hunk 2]
```diff
@@ -93,17 +93,17 @@
...
+PASSWORD_HASHERS = (
+    'django.contrib.auth.hashers.MD5PasswordHasher',
+)
+
+DATABASES = dict()
+DATABASES['default'] = {
+    'ENGINE': 'django.db.backends.sqlite3',
+    'NAME': str(BASE_DIR / 'test_db.sqlite3'),
+    'CONN_MAX_AGE': 60
+}
```
Comment (MaximilianWagner) [inline embark/embark/settings.py:106]: Please revert this change

**Review by MaximilianWagner — CHANGES_REQUESTED** [review]
A short explanation what to do, to send own asgi messages would be helpful for others.
e.g. extend from WebsocketConsumer, which methods are necessary.
Good work otherwise.

File: embark/embark/consumers.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            "phase": "",
+        }
+
+    # this method is executed when the connection to the frontend is established
+    def connect(self):
+        # accept socket connection
+        self.accept()
+
+        # if file does not exist create it otherwise delete its content
+        open('/app/emba/log_1/emba_new.log', 'w+')
```
Comment (ashutoshsingh0223) [inline embark/embark/consumers.py:40]: Are we sure the path will always be this?

File: embark/embark/consumers.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    def receive(self, text_data=None, bytes_data=None):
+        pass
+
+    # called when websocket connection is closed TODO: implement connection close if necessary
+    def disconnect(self, close_code):
+        pass
+
+    # send data to frontend
+    def send_data(self):
+        self.send(json.dumps(self.status_msg, sort_keys=True))
```
Comment (ashutoshsingh0223) [inline embark/embark/consumers.py:55]: Any reason for sorting the keys?

File: embark/embark/consumers.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    # update our dict whenever a new module is being processed
+    def update_status(self, stream_item_list):
+        # progress percentage TODO: improve percentage calculation
+        self.module_count += 1
+        percentage = self.module_count / 35
```
Comment (ashutoshsingh0223) [inline embark/embark/consumers.py:61]: Is 35 the total number of modules? If yes might better to add as constant in settings or env file

File: embark/embark/consumers.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        self.status_msg.update({"module": stream_item_list[0]})
```
Comment (ashutoshsingh0223) [inline embark/embark/consumers.py:62]: `self.status_msg["module"] = stream_item_list[0]`

This will do what you want. `update` is an overkill in my opinion

File: embark/embark/consumers.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        self.status_msg.update({"percentage": percentage})
```
Comment (ashutoshsingh0223) [inline embark/embark/consumers.py:63]: Same as above

File: embark/embark/consumers.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+                        self.input_processing(tmp)
+                        # copy diff to tmp file
+                        self.copy_file_content(tmp)
+
+    """
+              Regex function for lambda
+              :param inp: String to apply regex to
+              :param pat: Regex pattern
+              :return: True if regex matches otherwise False
+    """
```
Comment (ashutoshsingh0223) [inline embark/embark/consumers.py:99]: Doc strings should be inside the function just after definition

Comment (ashutoshsingh0223) [reply embark/embark/consumers.py:99]: Same recommendation for all doc strings

File: embark/embark/inotify_wrap.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+from inotify_simple import INotify, flags
+
+
+# wrapper for inotify system call -> get notification on file change
+def inotify_events():
+    inotify = INotify()
+    # TODO: add/remove flags to watch
+    watch_flags = flags.CREATE | flags.DELETE | flags.MODIFY | flags.DELETE_SELF | flags.CLOSE_NOWRITE | flags.CLOSE_WRITE
+    # add watch on file
+    inotify.add_watch('/app/emba/log_1/emba.log', watch_flags)
```
Comment (ashutoshsingh0223) [inline embark/embark/inotify_wrap.py:11]: Regarding file path are we sure we won't to get it from environment location or something?

File: embark/requirements.txt [hunk 10]
```diff
@@ -3,6 +3,12 @@ uWSGI
 ipython
 django-dotenv
 pycodestyle
+channels
+rx
+daphne
+inotify_simple
+inotify-simple
```
Comment (ashutoshsingh0223) [inline embark/requirements.txt:10]: Both are needed? What is the difference?

File: embark/requirements.txt [hunk 11]
```diff
@@ -3,6 +3,12 @@ uWSGI
...
+hypercorn
```
[the remark this answers is not in the document]
Comment (diegiesskanne) [reply embark/requirements.txt:11]: removed hypercorn

File: embark/requirements.txt (hunk 10 above)
Comment (diegiesskanne) [reply embark/requirements.txt:10]: im not sure. the documentation page says with '_' but in pycharm you only can install it with '-'. Is it a problem to keep them both?

File: embark/embark/consumers.py (hunk 4 above)
Comment (diegiesskanne) [reply embark/embark/consumers.py:55]: no its a mistake i think. i change it

File: embark/embark/consumers.py (hunk 3 above)
Comment (diegiesskanne) [reply embark/embark/consumers.py:40]: no its a placeholderfor testing purposes , but will be adjusted before the midterm release.

File: embark/embark/inotify_wrap.py (hunk 9 above)
Comment (diegiesskanne) [reply embark/embark/inotify_wrap.py:11]: no its a placeholderfor testing purposes , but will be adjusted before the midterm release.

File: embark/embark/consumers.py (hunk 5 above)
Comment (diegiesskanne) [reply embark/embark/consumers.py:61]: we can do that in the future
